api/experimental/arrow/v1/arrow_service.pb.go

```diff
@@ -249,6 +249,8 @@ type BatchArrowRecords struct {
 	ArrowPayloads []*ArrowPayload `protobuf:"bytes,2,rep,name=arrow_payloads,json=arrowPayloads,proto3" json:"arrow_payloads,omitempty"`
 	// [optional] Headers associated with this batch, encoded using hpack.
 	Headers []byte `protobuf:"bytes,3,opt,name=headers,proto3" json:"headers,omitempty"`
+	// [mandatory] The uncompressed size of the batch.
+	UncompressedSize uint64 `protobuf:"varint,4,opt,name=uncompressed_size,json=uncompressedSize,proto3" json:"uncompressed_size,omitempty"`
```

Comment:
I think we probably want to keep this information in a header value, since this uncompressed size corresponds with a specific non-arrow representation of the data. If there is Arrow data arriving from a pipeline that constructed Arrow data without an OTLP representation wouldn't know what to put here, and in that case we might need the receiver to fall back on some kind of estimate derived from somewhere else.

Therefore, I think we want to put the same information into the gRPC `metadata.MD` object on the exporter side, using ascii/decimal representation and let the header key be somehow tied to the OTel-Collector based in Go using OTLP pdata objects etc: I suggest `otlp-pdata-size`.

As a fallback (for missing uncompressed size), we can also unmarshal the object and afterwards acquire the decompressed-bytes difference. In case an SDK or some other pipeline produces data where there is not originally pdata. The problem with this approach is that the waiter is already holding the bytes that it is waiting for -- it suggests a new API may be needed to say "requesting more bytes, having acquired so many already".

This could lead to a scenario in which everyone is waiting, so requires care. The request to wait while holding acquired bytes has to (I think) release the bytes it holds while it waits for the total. As described, I think the worst-case memory held by waiters will be equal to the maximum waiter count times the maximum per-request byte limit. (That sounds good enough, to me!)

Reply:
(this comment really belongs here, https://github.com/open-telemetry/otel-arrow/pull/181/files#diff-6400e110ad8077bb46e7e66b00000c0adc011c5611c409be20e9d3622a28fc90 -- the idea is to avoid changing the protocol for the support in this PR)

Reply:
>> If there is Arrow data arriving from a pipeline that constructed Arrow data without an OTLP representation, it wouldn't know what to put here

I think I get your overall point here and am using your suggestion, but can you give an example of when this would happen? Shouldn't an otelarrowreceiver only accept requests that have an otlp representation? Is this mainly for testing the arrow stream?

Reply:
>> If there is Arrow data arriving from a pipeline that constructed Arrow data without an OTLP representation, it wouldn't know what to put here

I think I get your overall point here and am using your suggestion, but can you give an example of when this would happen? Shouldn't an otelarrowreceiver only accept requests that have an otlp representation? Is this mainly for testing the arrow stream?